Remotion/Data/DomainObjects.UnitTests/Infrastructure/ObjectIDStringSerialization/ObjectIDStringSerializerTest.cs

```diff
@@ -277,6 +277,44 @@ public void Parse_WithInvalidClassID_ThrowsFormatException ()
               "Serialized ObjectID 'Arder|5d09030c-25c2-4735-b514-46333bd28ac8|System.Guid' is invalid: 'Arder' is not a valid class ID."));
     }
 
+    [TestCase("Official|5d09030c-25c2-4735-b514-46333bd28ac8|System.Guid, mscorlib", typeof(Guid))]
+    [TestCase("Official|5d09030c-25c2-4735-b514-46333bd28ac8|System.Guid, mscorlib, Version=4.0.0.0", typeof(Guid))]
+    [TestCase("Official|23|System.Int32, mscorlib", typeof(Int32))]
+    [TestCase("Official|23|System.Int32 , mscorlib , Version=4.0.0.0", typeof(Int32))]
+    [TestCase("Official|test|System.String, mscorlib", typeof(String))]
+    [TestCase("Official|test| System.String , mscorlib , Version=4.0.0.0", typeof(String))]
+    [Test]
+    public void Parse_ExplicitAssemblyName (string idString, Type type)
+    {
+#if NET
+      Assert.That(
+          () => ObjectIDStringSerializer.Instance.Parse(idString),
+          Throws.TypeOf<FormatException>().With.Message.EqualTo(
+              $"Serialized ObjectID '{idString}' is invalid: type '{idString[(idString.LastIndexOf('|') + 1)..]}' is not supported."));
+#else
+      ObjectID id = ObjectIDStringSerializer.Instance.Parse(idString);
+
+      Assert.That(id.Value.GetType(), Is.EqualTo(type));
+#endif
+    }
+
+    [TestCase("Official|5d09030c-25c2-4735-b514-46333bd28ac8|System.Goid, mscorlib")]
+    [TestCase("Official|5d09030c-25c2-4735-b514-46333bd28ac8|System.guid, mscorlib")]
+    [TestCase("Official|5d09030c-25c2-4735-b514-46333bd28ac8|System.Guid, badassembly, mscorlib")]
+    [TestCase("Official|5d09030c-25c2-4735-b514-46333bd28ac8|System.Guid, badassembly, mscorlib, Version=4.0.0.0")]
+    [TestCase("Official|23|System.Int32, badassembly, mscorlib")]
+    [TestCase("Official|23|System.Int32 , badassembly, mscorlib , Version=4.0.0.0")]
+    [TestCase("Official|test|System.String, badassembly, mscorlib")]
+    [TestCase("Official|test| System.String , badassembly, mscorlib , Version=4.0.0.0")]
+    [Test]
+    public void Parse_WithInvalidAssemblyName_ThrowsFormatException (string idString)
+    {
+      Assert.That(
+          () => ObjectIDStringSerializer.Instance.Parse(idString),
+          Throws.TypeOf<FormatException>().With.Message.EqualTo(
+              $"Serialized ObjectID '{idString}' is invalid: type '{idString.Substring(idString.LastIndexOf('|') + 1)}' is not supported."));
```

Comment:
For unit tests, this sort of magic isn't a good thing (no, I don't want to discuss the rendering unit tests). Whenever possible, I do try to avoid arithemtic in the tests. We could simply add a parameter that represents the parsed typename.